csrc/device_lower/utils.cpp

```diff
@@ -1013,6 +1016,552 @@ bool predicateAtEnd(ForLoop* loop) {
+// linear in I1, and there is a 64 on the inner of [2] in I1. Because the I1 in
+// v1 and v3 are mapped, then we will continue propagation in v3. The next step
+// is to process the split of I1 into I1o and 256 in v3. We already know that
+// [2] is linear in I1, and on the inner of [2], there is a 64. Because 2 * 64 =
+// 128, which is a factor of 256, we know that the 256 is able to fully cover
+// the [2] in I1. Therefore, I1o is unrelated to [2], and we only need to focus
+// on the 256. And in this 256, there is a 64 on the inner of [2]. The next step
+// is to process the split of 256 into 4 and 64. Because the 64 happens to be
+// the size of the inner of [2], we know that the [2] will be fully covered by
+// the inner 2 of the 4 of the output of the split. Now we have finished
+// propagation and reached domain. We already proved that [2] is linear in
+// domain. In domain, the domains on the right of the 4 are 8, 8, 8, so the
+// stride is 8*8*8 = 512.
+namespace {
+
+// From the above example, we can see that how linear_g lives in domain could be
+// complicated. It can be, for example:
+//   1. linear_g is equivalent to a single ValGroup in domain.
+//   2. linear_g is the inner of a ValGroup in domain.
+//   3. linear_g is the outer of a ValGroup in domain.
+//   4. linear_g is the middle of a ValGroup in domain, where on the right,
+//      there is a 2.
+//   5. linear_g is split as g1, g2, g3 in domain.
+//   6. linear_g is split as the inner 2 of g1, g2, and the outer 4 of g3.
+//
+// We use a dynamic type to be able to represent all these cases.
+template <typename AbstractValGroup>
+struct PartOf {
+  std::shared_ptr<AbstractValGroup> group;
+  Val* inner_extent = nullptr;
+  Val* selected_extent = nullptr;
+};
+
+using AbstractValGroup = dynamic_type::DynamicType<
+    dynamic_type::Containers<
+        std::deque, // a composition of AbstractValGroups
+        PartOf // part of a AbstractValGroup
+        >,
+    ValGroup // a whole ValGroup
+    >;
+
+// Utilities to print AbstractValGroup.
+std::string print(const AbstractValGroup& group);
+
+std::string print(const ValGroup& group) {
+  return group->toString();
+}
+
+std::string print(const PartOf<AbstractValGroup>& part) {
+  auto str_or_null = [](Val* val) {
+    return val == nullptr ? "nullptr" : val->toInlineString();
+  };
+  return "PartOf(group=" + print(*part.group) +
+      ", inner_extent=" + str_or_null(part.inner_extent) +
+      ", selected_extent=" + str_or_null(part.selected_extent) + ")";
+}
+
+std::string print(const std::deque<AbstractValGroup>& vec) {
+  std::stringstream ss;
+  ss << "[";
+  for (const auto& g : vec) {
+    ss << print(g) << ", ";
+  }
+  ss << "]";
+  return ss.str();
+}
+
+std::string print(const std::monostate&) {
+  return "std::monostate";
+}
+
+std::string print(const AbstractValGroup& group) {
+  return AbstractValGroup::dispatch(
+      [&](const auto& group) { return print(group); }, group);
+}
```

Comment:
Should these be `toString` instead of `print`?

Reply:
renamed

Reply:
No, I have to change it back. `toString` is too heavily overloaded, and C++ have trouble finding the correct one.